pkg/hub/controllers/finalizercontroller/finalize_controller.go

```diff
@@ -0,0 +1,245 @@
+) factory.Controller {
+
+	controller := &FinalizeController{
+		roleLister:         roleInformer.Lister(),
+		roleBindingLister:  roleBindingInformer.Lister(),
+		namespaceLister:    namespaceLister,
+		manifestWorkClient: manifestWorkClient,
+		manifestWorkLister: manifestWorkLister,
+		rbacClient:         rbacClient,
+		eventRecorder:      eventRecorder,
+	}
+
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			key, _ := cache.MetaNamespaceKeyFunc(obj)
+			return key
+		}, roleInformer.Informer(), roleBindingInformer.Informer()).
+		WithSync(controller.sync).ToController("FinalizeController", eventRecorder)
+}
+
+func (m *FinalizeController) sync(ctx context.Context, controllerContext factory.SyncContext) error {
+	key := controllerContext.QueueKey()
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		// ignore role/rolebinding whose key is not in format: namespace/name
+		return nil
+	}
+
+	ns, err := m.namespaceLister.Get(namespace)
+	if err != nil {
+		return err
+	}
+
+	role, rolebinding, err := m.getRoleAndRoleBinding(namespace, name)
+	if err != nil {
+		return err
+	}
+
+	err = m.syncRoleAndRoleBinding(ctx, controllerContext, role, rolebinding, ns)
+
+	if err != nil {
+		klog.Errorf("Reconcile role/rolebinding %s fails with err: %v", key, err)
+	}
+	return err
+}
+
+func (m *FinalizeController) syncRoleAndRoleBinding(ctx context.Context, controllerContext factory.SyncContext,
+	role *rbacv1.Role, rolebinding *rbacv1.RoleBinding, ns *corev1.Namespace) error {
+	key := controllerContext.QueueKey()
+	klog.V(4).Infof("Sync role/rolebinding %s", key)
+
+	// no work if neither role nor rolebinding has the finalizer
+	if !hasFinalizer(role, manifestWorkFinalizer) && !hasFinalizer(rolebinding, manifestWorkFinalizer) {
+		return nil
+	}
+
+	if terminated(ns) {
+		works, err := m.manifestWorkLister.ManifestWorks(ns.Name).List(labels.Everything())
+		if err != nil {
+			return err
+		}
+
+		// requeue the key if there exists any manifest work in cluster namespace
+		if len(works) != 0 {
+			controllerContext.Queue().AddAfter(key, RequeueDelay)
+			klog.V(4).Infof("Requeue role/rolebinding %q after %v", key, RequeueDelay)
+			return nil
+		}
+
+		// remove finalizer from role/rolebinding
+		if err := m.removeFinalizerFromRole(ctx, role, manifestWorkFinalizer); err != nil {
+			return err
+		}
+
+		return m.removeFinalizerFromRoleBinding(ctx, rolebinding, manifestWorkFinalizer)
+	}
+
+	if terminated(role) {
+		if err := m.removeFinalizerFromRole(ctx, role, manifestWorkFinalizer); err != nil {
+			return err
+		}
+	}
+
+	if terminated(rolebinding) {
+		if err := m.removeFinalizerFromRoleBinding(ctx, rolebinding, manifestWorkFinalizer); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *FinalizeController) getRoleAndRoleBinding(namespace, name string) (*rbacv1.Role, *rbacv1.RoleBinding, error) {
+	role, err := m.roleLister.Roles(namespace).Get(name)
```

Comment:
i'm thinking whether we should filter the role by rules, so we only add finalizer to the roles who have rules to update/updatestatus of manifestworks?